File: Arrays/array-problem/121-best-time-to-buy-and-sell-stock.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

int maxProfit(vector<int>prices) {
    int bestBuy = prices[0];
    int maxProfit = 0;
    int n = prices.size();

    for(int i=1; i<n; i++) {
        if(prices[i] > bestBuy) {
            maxProfit = max(maxProfit, prices[i] - bestBuy);
        }
        bestBuy = min(bestBuy, prices[i]);
    }
    return maxProfit;
}

int main() {
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    int ans = maxProfit(prices);
    cout<< ans;
}

File: Arrays/array-problem/121-best-time-to-buy-and-sell-stock.js
const maxProfit = (prices) => {
  const n = prices.length;
  let maxProfit = 0;
  let bestBuy = prices[0];

  for (let i = 1; i < n; i++) {
    if (prices[i] > bestBuy) {
      maxProfit = Math.max(maxProfit, prices[i] - bestBuy);
    }
    bestBuy = Math.min(bestBuy, prices[i]);
  }
  return maxProfit;
};


function main() {
    const prices = [7, 2, 5, 9, 6, 4];
    ans = maxProfit(prices);
    console.log(ans); // 7
}

main();